rust/frontend/src/value.rs

```diff
@@ -0,0 +1,540 @@
+        let inner = ts::TVMValue { v_type: ty.inner };
+        Self::new(ValueKind::Type, inner)
+    }
+}
+
+impl<'a> From<&'a mut TVMType> for TVMValue {
+    fn from(ty: &mut TVMType) -> Self {
+        let inner = ts::TVMValue { v_type: ty.inner };
+        Self::new(ValueKind::Type, inner)
+    }
+}
+
+impl<'a> From<&'a TVMContext> for TVMValue {
+    fn from(ctx: &TVMContext) -> Self {
+        let inner = ts::TVMValue {
+            v_ctx: ctx.clone().into(),
+        };
+        Self::new(ValueKind::Context, inner)
+    }
+}
+
+impl<'a> From<&'a mut TVMContext> for TVMValue {
+    fn from(ctx: &mut TVMContext) -> Self {
+        let inner = ts::TVMValue {
+            v_ctx: ctx.clone().into(),
+        };
+        Self::new(ValueKind::Context, inner)
+    }
+}
+
+impl<'a> From<&'a TVMDeviceType> for TVMValue {
+    fn from(dev: &TVMDeviceType) -> Self {
+        let inner = ts::TVMValue {
+            v_int64: dev.0 as i64,
+        };
+        Self::new(ValueKind::Int, inner)
+    }
+}
+
+impl<'a> From<&'a mut TVMDeviceType> for TVMValue {
+    fn from(dev: &mut TVMDeviceType) -> Self {
+        let inner = ts::TVMValue {
+            v_int64: dev.0 as i64,
+        };
+        Self::new(ValueKind::Int, inner)
+    }
+}
+
+impl<'a> From<&'a TVMByteArray> for TVMValue {
+    fn from(barr: &TVMByteArray) -> Self {
+        let inner = ts::TVMValue {
+            v_handle: &barr.inner as *const ts::TVMByteArray as *mut c_void,
+        };
+        Self::new(ValueKind::Bytes, inner)
+    }
+}
+
+impl<'a> From<&'a mut TVMByteArray> for TVMValue {
+    fn from(barr: &mut TVMByteArray) -> Self {
+        let inner = ts::TVMValue {
+            v_handle: &barr.inner as *const ts::TVMByteArray as *mut c_void,
+        };
+        Self::new(ValueKind::Bytes, inner)
+    }
+}
+
+impl Debug for TVMValue {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        unsafe {
+            write!(
+                f,
+                "TVMValue: [v_int64: {:?}], [v_float64: {:?}], [v_handle: {:?}],\
+                 [v_str: {:?}]",
+                self.inner.v_int64, self.inner.v_float64, self.inner.v_handle, self.inner.v_str
+            )
+        }
+    }
+}
+
+impl Deref for TVMValue {
+    type Target = ts::TVMValue;
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl DerefMut for TVMValue {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+/// This type is needed for passing supported values as arguments to
+/// a [`function::Builder`]. Checkout the methods and from conversions.
+///
+/// ## Example
+///
+/// ```
+/// let ctx = TVMContext::gpu(0);
+/// let arg = TVMArgValue::from(&ctx);
+/// assert_eq!(arg.to_ctx(), ctx);
+/// ```
+///
+/// [`function::Builder`]:../function/struct.Builder.html
+#[derive(Debug, Clone)]
+pub struct TVMArgValue<'a> {
+    pub value: TVMValue,
+    pub type_code: TypeCode,
+    _lifetime: PhantomData<&'a ()>,
+}
+
+impl<'a> TVMArgValue<'a> {
```

Comment:
I suggest making a `common` crate and use the [runtime implementation of `TVMArgValue`](https://github.com/dmlc/tvm/blob/master/rust/src/runtime/packed_func.rs#L31)

Reply:
No can do, sorry. This is sufficiently fundamental to the architecture of Rust TVM that it's important to address now. The alternative would be having two different implementations of the same API which is confusing and bug-prone.

Don't worry, this will be solved soon! It'll be worth it; I promise :)

Reply:
Ok then, I guess that'd incur other changes such as unifying `TVMValue` and `TypeCode` which is more work than it seems!

Reply:
do feel free to pick away at any low-hanging fruits :)

Reply:
right! Are you currently working on this? 
I think, my `TVMValue` is more restrictive and I need it for debugging etc. The would cause some more changes in runtime.

Reply:
The low-hanging fruit now is that as the first step I can add `box_value: Box<Any>` and leave it as `box_value: box ()` and never use to because explicit `to_*` methods are enough to the conversions. Then merge this v0.1 PR. Later for next iteration we make a succinct plan for further unifications including 1) a common, 2) adding a builder to runtime and 3) make unified `type packedFunc=fn(&[TVMArgValue]) -> Result<TVMRetValue> for runtime.

What do you think?

Reply:
@nhynes ping

Reply:
I'm going to have to "strong no" on checking in code now with the intention of "fixing it later" because that rarely ever happens. If the frontend and runtime are truly so out-of-sync that a common crate can't be used for basic data types like `TVMPODValue`, then we have a larger design issue that needs to be solved.

Unless I'm mistaken (in which case you should please correct me), the crate which you're already maintaining is sufficient for users while we get the merge into TVM ready.

I'm sure that you'd prefer not to modify the frontend to use the runtime's `TVMPODValue` implementation, but I believe that it's the more faithful implementation of the C++ specification w.r.t. the internal representation and automatic conversion. As ever, I'd be glad to hear your perspective on this.

Reply:
@nhynes Thanks for your feedbacks! I meant that in the context of writing a plan to catch up the master version v0.5. I think going forward with the merge wouldn't cause any harm because 

1) `TVMArgValue` and `TVMRetValue` have the *same* interface now (already included `TVMRetValue` with `Box<Any>`)
2) I'm personally motivated to continue and voluntarily developing the frontend on my own, on my free time should show that 
3) I actually tried to find a common crate but it seems to me that runtime didn't necessarily consider debugging option which is a *must* for frontend, therefore the common crate would incur some good number of changes in runtime which would be too much for this PR in my opinion
4) My proposed future plans for Rust after this PR would be;
  i. Edition 2018
  ii. Finding a common crate and the changes it forces
  iii. Adding builder to runtime
  iv. Unifiying the PackedFunc as `fn(&[TVMArgValue]) -> Result<TVMRetValue>`

I hope I have provided enough evidence and motivations for keeping the ball rolling :)
Please let me know what would you think.

Reply:
@nhynes Alternatively, I can restart working on this in January and try to address the main common crate first, then if there won't be any other suggestions we move forward with merging this. Later plan for other proposed enhancements. Sounds good?